### PR TITLE
Implement escapeEscape helpers

### PR DESCRIPTION
**Pull request by Lillecarl**
Implements functions that escapes Terraform escape codes so that they're interpreted as strings when parsed through config.tf.json. This includes mapStringsDeep which applies a function on all strings in an attrset recursively through both attrsets and lists. It also applies toString to any stringlike type so it can be used by Terraform.

Escapes both ${} and %{} for consistency.

Discussion: I think traversing resources, locals and data and applying toString recursively to any stringlike type might make sense, that way if you stick derivations into values they're turned into storepaths automatically, or is it too much magic?

However what's in the PR is something I'm using right now to apply k8s JSON from kubenix through the TF kubectl provider (Cilium CNI has ${BIN_PATH} in some argument list somewhere) and I don't have a hole in my foot yet.

The formatting changes are nixfmt-rfc-style, I added rec since mapStringsDep calls itself but I could move it to a let block, doesn't look like it'll hurt here though tbf.

TODO: Docs?

### CODE REVIEW

**Comment by Lillecarl** [comment]
I double checked and numbers and such aren't stringLike either.

@mdarocha do you think this belongs here?

File: core/helpers.nix [hunk 1]
```diff
@@ -1,25 +1,57 @@
...
     # easier way to create templatefile() invocations
     # if used to set a variable, remember to wrap in `ref`
     template =
-      { text ? ""
-      , source ? ""
-      , variables ? { }
+      {
+        text ? "",
+        source ? "",
+        variables ? { },
```
Comment (sshine) [inline core/helpers.nix:14]: This reformatting doesn't contribute to the purpose of the pull request. A downside of formatting for formatting's sake is that it clutters the version history:
- the most recent history of reformatted lines are no longer tied to the logical changes that introduced them
- the logical changes of this pull request are harder to read

Since you picked nixfmt-rfc-style, it could lead towards having uniform formatting. But I think it's better to deal with this in a separate pull request.

File: core/helpers.nix [hunk 2]
```diff
@@ -1,25 +1,57 @@
...
-            else source;
-        in
-        "templatefile(\"${template}\", ${builtins.toJSON variables})";
+      assert super.assertMsg (
+        text == "" || source == ""
+      ) "You must provide either 'text' or 'source' arguments";
+      let
+        template = if text != "" then pkgs.writeText "template.tftpl" text else source;
+      in
+      "templatefile(\"${template}\", ${builtins.toJSON variables})";
```
Comment (sshine) [inline core/helpers.nix:22]: It appears that there are no logical changes here, either.

File: core/helpers.nix [hunk 3]
```diff
@@ -1,25 +1,57 @@
...
+        function (toString value)
+      else if super.isAttrs value then
+        super.mapAttrs (name: mapStringsDeep function) value
+      else if super.isList value then
+        map (mapStringsDeep function) value
+      else
+        value;
+
+    # Escapes strings so variables aren't interpolated with ${} and %{}
+    escapeEscape =
```
Comment (sshine) [inline core/helpers.nix:38]: Shouldn't this one be called `escapeInterpolation`?

The ["escape escape" sequences](https://developer.hashicorp.com/terraform/language/expressions/strings#escape-sequences) are `$${}` and `%%{}`

```suggestion
    escapeInterpolation =
```

File: core/helpers.nix [hunk 4]
```diff
@@ -1,25 +1,57 @@
 # override to the standard nixpkgs lib, adding
 # terranix-related helpers
 pkgs: self: super: {
-  tf = {
+  tf = rec {
```
Comment (sshine) [inline core/helpers.nix:4]: I'd prefer to avoid it as a principle, but others may chime in and disagree.

**Review by sshine — COMMENTED** [review]
Hello, and thanks for the pull request!

I should start by saying I'm not qualified to merge this at this point in time, since it's a little far for me to assess it. I have these concerns:

1. Is this the right approach? (Escaping is tricky to get right.)
2. Does it address escaping `%%{}` and `$${}` properly?
3. Can we test this?

I think adding documentation for any improvement could be nice, but it might be discovered by those who need it. More importantly, I'd like if it were tested, since I sense there are some corner cases.

I won't approve or reject anything for now. I can come back and assess when I am closer to being a user of this, or I thoroughly understand the consequences of applying escaping like this.

Alternatively, @mdarocha is welcome to make their say.

**Comment by Lillecarl** [comment]
I appreciate you taking time to review, but style feedback from non-maintainers on month-old PRs isn't particularly helpful. If maintainers want changes, they can request them.

**Review by sshine — CHANGES_REQUESTED** [review]
Yeah, it'd be weird if I made those requests and I didn't have agency to merge. :-D I was a maintainer since 2025-04-25 in order to prevent the project from being labelled as unmaintained in spite of @mdarocha's efforts:

- #122

While the original BDFL is lurking in the background for now, neither @mdarocha nor I have absolute authority on where to take the project, and there is no clear alignment (we may approve things without consolidating).

Addressing the unnecessary whitespace changes: This is not an argument about style, but of maintainability. I could be in favor of making the whole codebase conform to nixfmt-rfc-style, but I'd like to do that separate from this. This is my only request; as for the escaping itself, I will defer from evaluating it until I come around and understand the implications better, since this is untested and does not analyse or touch corner cases.

**Comment by Lillecarl** [comment]
@sshine I've glanced at the tests and tried to implement nr 15 but I got some odd bash errors, I'll see when I can squeeze it in.
